virtual_rainforest/abiotic/radiation.py

```diff
@@ -0,0 +1,179 @@
+
+# the following structural components are not implemented yet
+TODO core.constants
+TODO include vertical structure (canopy layers)
+TODO include data structure (->numpy array)
+TODO include time dimension
+TODO logging, raise errors
+"""
+
+# import numpy as np
+# from core.constants import kc, kd, kfFEC, bolz # this doesn't exist yet
+kc = 1
+kd = 1
+kfFEC = 1
+bolz = 1
+
+
+class Radiation:
+    """Radiation balance.
+
+    Attributes:
+        topofcanopy_radiation: float, shortwave radiation, top of canopy [W m-2]
+        ppfd: float, photosynth. photon flux density, top of canopy [mol m-2]
+        rn_profile: float, net radiation profile, below canopy [W m-2]
+    """
+
+    def __init__(self) -> None:
+        """Initializes point-based radiation method."""
+        pass
+
+    def calc_sw_down(self, n: int, y: int, lat: float) -> float:
+        """Calculate shortwave downward radiation.
+
+        This function calculated incoming shortwave radiation based on location and time
+        of year if this is NOT provided as an input to the virtual rainforest.
+
+        Args:
+            n: day of the year
+            y: year
+            lat: latitude
+
+        Ref:
+            Davis et al. (2017): Simple process-led algorithms for simulating habitats
+            (SPLASH v.1.0): robust indices of radiation, evapotranspiration and plant-
+            available moisture, Geosci. Model Dev., 10, 689-708
+        """
+        raise NotImplementedError("Implementation of this feature is missing")
+
+    def calc_topofcanopy_radiation(
+        self,
+        sw_in: float,
+        n: int,
+        y: int,
+        lat: float,  # this is optional for the case of sw_in == FALSE
+        elev: float,
+        sf: float,
+        topo_factor: float,
+        sw_albedo: float,
+        vis_albedo: float,
+    ) -> None:
+        """Calculate top of canopy shortwave radiation.
+
+        If sw_in (extra-terrestrial) radiation is NOT provided as an input, it is
+        calculated as a function of latitude and date based on the SPLASH model
+        (Davis et al., 2017).
+
+        Args:
+            sw_in: shortwave downward radiation [W m-2]
+            n: day of year
+            y: year
+            lat: latitude
+            elev: elevation (m)
+            sf: fraction of sunshine hours
+            topo_factor: topographic adjustment factor for radiation (slope and aspect)
+            sw_albedo: shortwave albedo (function of vegetation and bare soil fraction)
+            vis_albedo: albedo visible light (for PPFD calculation)
+
+        Returns:
+            topofcanopy_radiation
+            ppfd
+        """
+        # check if shortwave down is provided, else calculate
+        if sw_in:
+            sw_down = sw_in
+        else:
+            sw_down = self.calc_sw_down(n, y, lat)
+
+        # atmospheric filtering (clouds, aerosols)
+        tau_o = kc + kd * sf
+        tau = tau_o * (1.0 + (2.67e-5) * elev)
```

Comment:
Should the `2.67e-5` be part of the constants?

Reply:
This is not a universal constant, the equation is based on the regression of Beer’s radiation extinction function at elevations below 3000 m, so I think it could stay like thi,s maybe best to add the reference?

Reply:
Yes - but then if someone publishes a new version, you don't want to have to change the hard-coding and people might want to try different versions. If this is is coded as part of a `constants` library then it can be changed flexibly.

Reply:
There are likely better ways to do this, but `pyrealm` has basically no hard-coded parameters in it at all and has sets of lookup tables to store things like this:
https://github.com/davidorme/pyrealm/blob/develop/pyrealm/param_classes.py

Reply:
@vgro Even if you don't make this a global constant, you could make it a local one. I'd personally make it caps to show it's a constant and then just define it at the start of the function, maybe with a docstring saying what it is:

```py
BEER_REGRESSION = 2.67e-5
"""Parameter which does something that Alex doesn't understand."""
```

Reply:
Some bigger picture considerations:

1. Over in `pyrealm` I want to be able to have parallel instances of a PModel that might differ in these 'constant' parameters. That is the _main_ reason they are bundled up in classes: each PModel instance can have a separate isolated instance of the parameters. If they were declared as global constants, you could still change them but it would get extremely error prone. I don't think we want that here - but it is a big part of why that example is structured as it is.

2. But we do have an existing similar system: this  _could_ be moved into `config`? Part of the point of that system is to fill in defaults when users don't want to tweak them, so it could be declared there.

Reply:
@davidorme I agree that we might want to have different values for certain constants but I don't think this is one of them.

Reply:
I've certainly taken an extreme view over in `pyrealm` 😄 You could set it up to run on Mars or a parallel universe...

Reply:
> @alexdewar good pont, I'll do that. is it a general convention to have constants in caps?

Yep, though not everyone does it. I like that it makes them stand out when you're skimming the code.

Reply:
> I've certainly taken an extreme view over in `pyrealm` 😄 You could set it up to run on Mars or a parallel universe...

I agree it is good to have these sort of parameters in one place. I just had a closer look into the pyrealm code and it makes a lot of sense to me to structure it that way. I'll keep it as it is for now and we can merge it once we have the location for it :-)